service/raft/raft_group0_client.hh

```diff
@@ -190,4 +191,59 @@ public:
     void set_query_result(utils::UUID query_id, service::broadcast_tables::query_result qr);
 };
 
+using mutations_generator = coroutine::experimental::generator<mutation>;
+
+// mutations_collector is used to gather mutations which are side effects
+// of functions execution. They need to be announced under single guard
+// for atomicity. As functions which produce mutations may embed each other
+// we need to decouple announcing step to a common external place here.
+// It also supports generator callbacks to avoid holding too many mutations
+// in memory.
+//
+// There is no guarantee about ordering of mutations within singe transaction.
+// Single mutations_collector object represents a single transaction.
+// If size or number of mutations is too big for raft too handle it will be
+// rejected.
+class mutations_collector {
+public:
+    using generator_func = std::function<mutations_generator(api::timestamp_type t)>;
+private:
+    std::vector<mutation> _muts;
+    std::vector<generator_func> _generators;
+    std::optional<::service::group0_guard> _guard;
```

Comment:
Why is it optional?

Reply:
(commented in the consutructor) due to legacy v1 path which doesn't have the guard and we don't want to guard it by ifs in dozen places